xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -59,520 +67,763 @@ define(["jquery", "polyfills"], function ($) {
-  };
 
+    /**
+     * Get the page corresponding to the specified entry (0-based index)
+     * @param {Number} entryIndex The index of the entry. Uses current entry if undefined.
+     * @returns {Number}
+     */
+    getPageIndex: function (entryIndex) {
+      if (entryIndex === undefined) {
+        entryIndex = this.data.query.offset;
+      }
+      return Math.floor(entryIndex / this.data.query.limit);
+    },
+
+
+    /**
+     * Set page index (0-based index), then fetch new data
+     * @param {Number} pageIndex
+     * @returns {Number}
+     */
+    setPageIndex: function (pageIndex) {
+      if (pageIndex < 0 || pageIndex >= this.getPageCount()) { return; }
+      var previousPageIndex = this.data.query.offset;
+      this.data.query.offset = this.getFirstIndexOfPage(pageIndex);
+      this.triggerEvent("pageChange", {
+        pageIndex: pageIndex,
+        previousPageIndex: previousPageIndex,
+      });
+      // CALL FUNCTION TO FETCH NEW DATA HERE
+    },
+
+
+    /**
+     * Get the first entry index of the given page index
+     * @param {Number} pageIndex The page index. Uses current page if undefined.
+     * @returns {Number}
+     */
+    getFirstIndexOfPage: function (pageIndex) {
+      if (pageIndex === undefined) {
+        pageIndex = this.getPageIndex();
+      }
+      if (0 <= pageIndex && pageIndex < this.getPageCount()) {
+        return pageIndex * this.data.query.limit;
+      } else {
+        return -1;
+      }
+    },
 
-  /**
-   * Return a new displayer based on the specified property and row data
-   * @param {String} propertyId The id of the property of the entry
-   * @param {Object} entry The entry data object
-   * @param {String} displayerId
-   * @returns {Promise}
-   */
-  Logic.prototype.createDisplayer = function (propertyId, entry, displayerId) {
-    var self = this;
 
-    return new Promise (function (resolve, reject) {
-      // default displayerId
-      if (displayerId === undefined) {
-        displayerId = self.getDisplayerDescriptor(propertyId).id;
+    /**
+     * Get the last entry index of the given page index
+     * @param {Number} pageIndex The page index. Uses current page if undefined.
+     * @returns {Number}
+     */
+    getLastIndexOfPage: function (pageIndex) {
+      if (pageIndex === undefined) {
+        pageIndex = this.getPageIndex();
       }
+      if (0 <= pageIndex && pageIndex < this.getPageCount()) {
+        return Math.min(this.getFirstIndexOfPage(pageIndex) + this.data.query.limit, this.data.data.count) - 1;
+      } else {
+        return -1;
+      }
+    },
+
+
+    /**
+     * Set the pagination page size, then fetch new data
+     * @param {Number} pageSize
+     */
+    setPageSize: function (pageSize) {
+      if (pageSize < 0) { return; }
+      var previousPageSize = this.data.query.limit;
+      this.data.query.limit = pageSize;
+      this.triggerEvent("pageSizeChange", {
```

Comment:
You could also check if the new page size is different than the previous page size.